lang/java/reef-utils/src/main/java/org/apache/reef/util/MultiAsyncToSync.java

```diff
@@ -0,0 +1,105 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.reef.util;
+
+import java.util.ArrayDeque;
+import java.util.HashMap;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * Assists a class in presenting a synchronous interface that is implemented
+ * via asynchronous interfaces and events. When a method call is received
+ * by the interface, parameter values captured and asynchronous processing
+ * started, the caller is put to sleep by calling block() with the internal
+ * interface caller identifier. After all of the asynchronous processing
+ * is complete the caller is released with a call to release().
+ */
+public final class MultiAsyncToSync {
+  private static final Logger LOG = Logger.getLogger(MultiAsyncToSync.class.getName());
+
+  private final ArrayDeque<ConditionVariable> freeQueue = new ArrayDeque<>();
```

Comment:
There are a number of design issues involved here.
1) Allocating ConditionVaribles is expensive so it makes sense to reuse them, especially when message traffic is high. Some uses of the MultiAsyncToSync class, when no more than one call is in flight at one time, will continually move a single ConditionVariable object back and forth from the deque to the map. However, when multiple calls are in flight simultaneously, we will only allocate a new ConditionVariable when we reach a new high of calls in flight and we eventually reach a high water mark and allocation stops. 
2) The map is necessary so that we know who to wake up for a given message acknowledgment.
3) I chose non-concurrent containers because only one synchronization event is required to manage both containers. Choosing concurrent containers would cause two synchronization events to occur during the move from the free queue to the map and visa versa. Synchronization events tend to be very expensive due to potential cache flushes.
4) I would prefer to use a simple queue here but that class is abstract in Java.